DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/mmhost/README.md

```diff
@@ -0,0 +1,372 @@
+
+#### Example 1: A kernel with multiple pointer arguments
+(Code can be found under `part1_pointers`)
+```c++
+struct PointerIP {
+  // Pointer kernel arguments will be passed through the component's CSR. They
+  // will refer to data accessible through a shared Avalon memory-mapped host
+  // interface.
+  int *x;
+  int *y;
+  int *z;
+  int size;
+
+  void operator()() const {
+    for (int i = 0; i < size; ++i) {
+      z[i] = x[i] + y[i];
+    }
+  }
+};
+```
+![](assets/visio-naive.svg)
+
+### Using `annotated_arg` to configure Avalon memory-mapped host interfaces
+
+The default behaviour of a pointer argument can be overridden by declaring an `annotated_arg` kernel argument.
+
+#### Example 2: A kernel with a single customized Avalon memory-mapped host interface
+(Code can be found under `part2_single_host`).
+```c++
+struct SingleMMIP {
+  // This kernel has 3 annotated pointers, but since they have no properties
+  // specified, this kernel will result in the same IP component as Example 1.
+  sycl::ext::oneapi::experimental::annotated_arg<int *> x;
+  sycl::ext::oneapi::experimental::annotated_arg<int *> y;
+  sycl::ext::oneapi::experimental::annotated_arg<int *> z;
+  int size;
+
+  void operator()() const {
+    for (int i = 0; i < size; ++i) {
+      z[i] = x[i] + y[i];
+    }
+  }
+};
+```
+
+The following table describes the properties under `sycl::ext::intel::experimental` that can be used to customize how the pointer argument is passed to the component. Only one may be specified at a time. 
+
+| Parameter                 | Description
+|---                        |---
+| `register_map`            | Pass the pointer for this memory-mapped host interface through the IP component's control/status register
+| `conduit`                 | Pass the pointer for this memory-mapped host interface through a conduit interface
+| `stable`                 | N/A | User guarantee that the pointer will not change between pipelined invocations of the kernel. The compiler uses this to further optimize the kernel.
+
+
+The following parameters are found under `sycl::ext::intel::experimental`, with the exception of `alignment` under `sycl::ext::oneapi::experimental`. They can be used to configure an IP component's Avalon memory-mapped host interfaces:
+
+| Parameter                | Default Value | Description
+|---                       |---            |---
+|  `buffer_location<id>`   | N/A           | The address space of the interface that associates with the host. Each unique buffer location will result in a unique Avalon memory-mapped host interface. When `buffer_location` is not specified, then corresponding LSUs will connect to all of the IP's Avalon memory-mapped host interfaces.
```

Comment:
I want to not say "LSU connects to all interfaces" - that is more like potential implementation detail. I think a user needs to know that if pointer to mem_X is passed via an unannotated pointer, then the pointer dereference will access the Avalon interface for mem_X.

It is a user error to perform read access on some mem_X with a raw pointer when mem_X interface is a write only interface (and similarly in analogous write case) - I think we can mention that here somewhere as a cautionary note.

Thoughts @whitepau ?


```suggestion
|  `buffer_location<id>`   | N/A           | The address space of the interface that associates with the host. Each unique buffer location will result in a unique Avalon memory-mapped host interface. When `buffer_location` is not specified, then the pointer can be used to access any of the IP's Avalon memory-mapped host interfaces depending on which memory the pointer points to.
```